hazelcast/src/main/java/com/hazelcast/cache/impl/CacheSplitBrainHandlerService.java

```diff
@@ -86,47 +76,63 @@
-        this.logger = nodeEngine.getLogger(getClass());
     }
 
     @Override
     public Runnable prepareMergeRunnable() {
-        Map<String, Map<Data, CacheRecord>> recordMap = createHashMap(configs.size());
+        Map<String, Collection<ICacheRecordStore>> collectedStores = new HashMap<String, Collection<ICacheRecordStore>>();
+        Map<String, Collection<ICacheRecordStore>> collectedStoresWithLegacyPolicies
+                = new HashMap<String, Collection<ICacheRecordStore>>();
+        List<ICacheRecordStore> backupStores = new LinkedList<ICacheRecordStore>();
 
         for (int partitionId = 0; partitionId < partitionCount; partitionId++) {
-            // add your owned entries so they will be merged
-            if (partitionService.isPartitionOwner(partitionId)) {
-                CachePartitionSegment segment = segments[partitionId];
-                List<Iterator<ICacheRecordStore>> iterators = iteratorsOf(segment);
-                for (Iterator<ICacheRecordStore> iterator : iterators) {
-                    while (iterator.hasNext()) {
-                        ICacheRecordStore cacheRecordStore = iterator.next();
-                        String cacheName = cacheRecordStore.getName();
-                        if (cacheRecordStore.getConfig().getInMemoryFormat() == NATIVE
-                                && nodeEngine.getClusterService().getClusterVersion().isLessThan(Versions.V3_10)) {
-                            logger.warning("Split-brain recovery can not be applied NATIVE in-memory-formatted cache ["
-                                    + cacheName + ']');
-                            continue;
-                        }
-
-                        Map<Data, CacheRecord> records = recordMap.get(cacheName);
-                        if (records == null) {
-                            records = createHashMap(cacheRecordStore.size());
-                            recordMap.put(cacheName, records);
-                        }
-                        for (Map.Entry<Data, CacheRecord> cacheRecordEntry : cacheRecordStore.getReadOnlyRecords().entrySet()) {
-                            Data key = cacheRecordEntry.getKey();
-                            CacheRecord cacheRecord = cacheRecordEntry.getValue();
-                            records.put(key, cacheRecord);
-                        }
+            boolean partitionOwner = partitionService.isPartitionOwner(partitionId);
+            CachePartitionSegment segment = segments[partitionId];
+            List<Iterator<ICacheRecordStore>> iterators = iteratorsOf(segment);
+            for (Iterator<ICacheRecordStore> iterator : iterators) {
+                while (iterator.hasNext()) {
+                    ICacheRecordStore cacheRecordStore = iterator.next();
+
+                    if (!partitionOwner) {
+                        backupStores.add(cacheRecordStore);
+                        continue;
+                    }
+
```

Comment:
I think we're missing the `DiscardMergePolicy` check here. I don't even find it in the current master, so I completely forgot it before, or it was removed during a refactoring. Nevertheless, we should put it (back) here.

Reply:
fixed